clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,130 @@
+          hasCondition(binaryOperator(hasAnyOperatorName("<", ">", "<=", ">="),
+                                      hasLHS(expr().bind("CondLhs")),
+                                      hasRHS(expr().bind("CondRhs")))
+                           .bind("binaryOp")),
+          hasThen(
+              anyOf(stmt(AssignOperator),
+                    compoundStmt(statementCountIs(1), has(AssignOperator)))),
+          hasParent(stmt(unless(ifStmt(hasElse(
+              equalsBoundNode("if"))))))), // Ensure `if` has no `else if`
+      this);
+}
+
+void UseStdMinMaxCheck::registerPPCallbacks(const SourceManager &SM,
+                                            Preprocessor *PP,
+                                            Preprocessor *ModuleExpanderPP) {
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void UseStdMinMaxCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *If = Result.Nodes.getNodeAs<IfStmt>("if");
+  const auto &Context = *Result.Context;
+  const auto &LO = Context.getLangOpts();
+  const SourceManager &Source = Context.getSourceManager();
+
+  const SourceLocation IfLocation = If->getIfLoc();
+  const SourceLocation ThenLocation = If->getEndLoc();
+
+  // Ignore Macros
+  if (IfLocation.isMacroID() || ThenLocation.isMacroID())
+    return;
+
+  const auto *CondLhs = Result.Nodes.getNodeAs<Expr>("CondLhs");
+  const auto *CondRhs = Result.Nodes.getNodeAs<Expr>("CondRhs");
+  const auto *AssignLhs = Result.Nodes.getNodeAs<Expr>("AssignLhs");
+  const auto *AssignRhs = Result.Nodes.getNodeAs<Expr>("AssignRhs");
+
+  const auto CreateReplacement = [&](bool UseMax) {
+    const auto *FunctionName = UseMax ? "std::max" : "std::min";
+    const auto CondLhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(CondLhs->getSourceRange()), Source, LO);
+    const auto CondRhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(CondRhs->getSourceRange()), Source, LO);
+    const auto AssignLhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(AssignLhs->getSourceRange()), Source, LO);
+    return (AssignLhsStr + " = " + FunctionName +
+            ((CondLhs->getType() != CondRhs->getType())
+                 ? "<" + AssignLhs->getType().getAsString() + ">("
+                 : "(") +
+            CondLhsStr + ", " + CondRhsStr + ");")
+        .str();
+  };
+  const auto *BinaryOp = Result.Nodes.getNodeAs<BinaryOperator>("binaryOp");
+  const auto BinaryOpcode = BinaryOp->getOpcode();
+  const auto OperatorStr = BinaryOp->getOpcodeStr();
+  if (((BinaryOpcode == BO_LT || BinaryOpcode == BO_LE) &&
+       (tidy::utils::areStatementsIdentical(CondLhs, AssignRhs, Context) &&
+        tidy::utils::areStatementsIdentical(CondRhs, AssignLhs, Context))) ||
+      ((BinaryOpcode == BO_GT || BinaryOpcode == BO_GE) &&
+       (tidy::utils::areStatementsIdentical(CondLhs, AssignLhs, Context) &&
+        tidy::utils::areStatementsIdentical(CondRhs, AssignRhs, Context)))) {
```

Comment:
consider extracting those 2 big if's conditions into separate static functions, same for CreateReplacement  lambda.
It would make check method smaller and therefore more readable.